[Inductor] fix broadcast logic for Triton

nandesuka opened the pull request:
Summary: Fix logic for inserting broadcast on kernel with load going directly to store. In the case where load is going directly to store, we insert a tl.broadcast on the store, regardless of the block size on the load. In the case where a broadcast is not required, the downstream Triton compiler is expected to remove this no-op broadcast instruction.

Test Plan: Added tests under test_torchinductor_strided_blocks.py:test_expand_broadcast in OSS and internal test cases.

Reviewed By: blaine-rister

Differential Revision: D65518033




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

nandesuka commented:
@pytorchbot label "topic: not user facing"

blaine-rister commented on test/inductor/test_torchinductor_strided_blocks.py:
```diff
@@ -229,8 +228,56 @@ def get_input(view_size: Tuple[int]) -> torch.Tensor:
+            y
+        )
+            
+    @parametrize("prefer_nd_tiling", [True])
```
Seems like this change was accidental? This used to be

`@parametrize("prefer_nd_tiling", [False, True])`

nandesuka replied:
Yup, good catch. Will update.

blaine-rister reviewed (approved):
LGTM, thanks for the fix. I left a comment about one of the tests being changed. Also it looks like the expected code for one of the CI tests needs to be updated. Once the CI is green, this looks ready to go :)

nandesuka committed:
[Inductor] fix broadcast logic for Triton (#141027)

Summary:

Fix logic for inserting broadcast on kernel with load going directly to store. In the case where load is going directly to store, we insert a tl.broadcast on the store, regardless of the block size on the load. In the case where a broadcast is not required, the downstream Triton compiler is expected to remove this no-op broadcast instruction.

Test Plan: Added tests under test_torchinductor_strided_blocks.py:test_expand_broadcast in OSS and internal test cases.

Reviewed By: blaine-rister

Differential Revision: D65518033

nandesuka commented:
@pytorchbot rebase
